debian/patches/0080-add-tonemapx-filter.patch

```diff
@@ -0,0 +1,3269 @@
++void ff_get_yuv_coeffs(int16_t out[3][3][8], double (*table)[3],
++                       int depth, int y_rng, int uv_rng, int yuv2rgb);
+ #endif
+Index: FFmpeg/libavfilter/Makefile
+===================================================================
+--- FFmpeg.orig/libavfilter/Makefile
++++ FFmpeg/libavfilter/Makefile
+@@ -516,6 +516,7 @@ OBJS-$(CONFIG_TMEDIAN_FILTER)
+ OBJS-$(CONFIG_TMIDEQUALIZER_FILTER)          += vf_tmidequalizer.o
+ OBJS-$(CONFIG_TMIX_FILTER)                   += vf_mix.o framesync.o
+ OBJS-$(CONFIG_TONEMAP_FILTER)                += vf_tonemap.o
++OBJS-$(CONFIG_TONEMAPX_FILTER)               += vf_tonemapx.o
+ OBJS-$(CONFIG_TONEMAP_OPENCL_FILTER)         += vf_tonemap_opencl.o opencl.o \
+                                                 opencl/tonemap.o opencl/colorspace_common.o
+ OBJS-$(CONFIG_TONEMAP_CUDA_FILTER)           += vf_tonemap_cuda.o cuda/tonemap.ptx.o \
+Index: FFmpeg/libavfilter/vf_tonemapx.c
+===================================================================
+--- /dev/null
++++ FFmpeg/libavfilter/vf_tonemapx.c
+@@ -0,0 +1,3151 @@
++/*
++ * This file is part of FFmpeg.
++ *
++ * FFmpeg is free software; you can redistribute it and/or
++ * modify it under the terms of the GNU Lesser General Public
++ * License as published by the Free Software Foundation; either
++ * version 2.1 of the License, or (at your option) any later version.
++ *
++ * FFmpeg is distributed in the hope that it will be useful,
++ * but WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++ * Lesser General Public License for more details.
++ *
++ * You should have received a copy of the GNU Lesser General Public
++ * License along with FFmpeg; if not, write to the Free Software
++ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
++ */
++
++/**
++ * @file
++ * tonemap algorithms
++ */
++
++#include <float.h>
++#include <stdio.h>
++#include <string.h>
++
++#include "libavutil/avassert.h"
++#include "libavutil/imgutils.h"
++#include "libavutil/internal.h"
++#include "libavutil/intreadwrite.h"
++#include "libavutil/mem_internal.h"
++#include "libavutil/opt.h"
++#include "libavutil/pixdesc.h"
++#include "libavutil/cpu.h"
++#if  ARCH_AARCH64
++#   include <arm_neon.h>
++#   include "libavutil/aarch64/cpu.h"
++#endif
++#if ARCH_X86
++#   include <immintrin.h>
++#   include <emmintrin.h>
++#   include <smmintrin.h>
++#   include "libavutil/x86/cpu.h"
++#endif
++
++#include "avfilter.h"
++#include "colorspace.h"
++#include "formats.h"
++#include "internal.h"
++#include "video.h"
++
++#define REFERENCE_WHITE 203.0f
++#define FLOAT_EPS 1.175494351e-38f
++
++#if defined(__GNUC__) || defined(__clang__)
++#    if (__GNUC__ >= 11) || (__clang_major__ >= 12)
++#        define X86_64_V2 __attribute__((target("arch=x86-64-v2")))
++#        define X86_64_V3 __attribute__((target("arch=x86-64-v3")))
```

Comment:
Still not a fan of targeting x96_64 feature levels while not checking all of the requirements at runtime

Reply:
It is widely used. Microsoft MSVC does not even have the feature level, and if you enable AVX2 it will enable AVX2+FMA+BMI1+BMI2

Reply:
One valid use case for not using feature level is to support "fake" CPUs, like the CPUs in a VM. The hypervisor can disable instruction extensions as they wish to make them behave differently from all real CPUs.

Reply:
Does AVX2+FMA3 support imply x86-64-v3 in real hardware?
If not, does targeting `avx2,fma` instead of `x86-64-v3` hurt performance?

Reply:
> Does AVX2+FMA3 support imply x86-64-v3 in real hardware?

For all real CPUs from Intel and AMD, yes, all CPUs from them with both AVX2 and FMA3 conforms to the x86-64-v3 feature level. The level feature level is decided with the real CPUs after all.

Reply:
@Bond-009 Do you have a consensus on this?

We actually use intrinsics up to SSE4.1 and AVX2+FMA. So conservatively speaking, if you don't want to use x86-64-v2/v3, we can use this instead:
```
__attribute__((target("sse4.2")))
__attribute__((target("avx2,fma")))
```
This is a subset of the former, but I haven't tested whether they have any impact on perf versus x86-64-v2/v3.

Reply:
@nyanmisaka I'm totally fine with targeting the specific ISA extensions. I'm just not a fan of using x86_64 feature levels without checking all requirements. Especially if it's for improved performance but no benchmark results are provided.

Reply:
> @nyanmisaka I'm totally fine with targeting the specific ISA extensions. I'm just not a fan of using x86_64 feature levels without checking all requirements. Especially if it's for improved performance but no benchmark results are provided.

@gnattu

Reply:
I did a benchmark and found no measurable perf difference between targeting ISA level and specific instructions. I pushed to always use instruction targets.